### PR TITLE
google cloud functions integration

### PR DESCRIPTION
**Pull request by orlichter1**
## Status
- [ ] In Progress
- [x] Ready (except proxy test)
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23147

## Description
A new google cloud functions integration.

## Minimum version of Demisto
- [ ] 4.5.0
- [x] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation 

Files for tech review:
- [ ] [YML](https://github.com/demisto/content/edit/Google_cloud_functions/Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F6391)
- [ ] [Description file](https://github.com/demisto/content/edit/Google_cloud_functions/Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_description.md?pr=%2Fdemisto%2Fcontent%2Fpull%2F6391)

### CODE REVIEW

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+- display: Default Region
+  name: region
+  required: false
+  type: 0
+- display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
+- display: Trust any certificate (not secure)
+  name: insecure
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:23]: this is not used in the code

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  required: false
+  type: 8
+description: Google Cloud Functions is an event-driven serverless compute platform.
+  Run your code locally or in the cloud without having to provision servers.
+display: Google Cloud Functions
+name: GoogleCloudFunctions
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:29]:
```suggestion
name: Google Cloud Functions
```

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml [hunk 3]
```diff
@@ -0,0 +1,165 @@
+category: Utilities
+commonfields:
+  id: GoogleCloudFunctions
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:3]:
```suggestion
  id: Google Cloud Functions
```

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    '''EXECUTION CODE'''
+    cmd_func = demisto.command()
+    LOG(f'Command being called is {cmd_func}')
+    try:
+        if cmd_func == 'test-module':
+            region_list_command(client, {})
+            demisto.results('ok')
+            sys.exit(0)
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:165]: remove sys.exit

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        if not cmd_func:
+            raise NotImplementedError(f'Command "{cmd_func}" is not implemented.')
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:167]: remove this, its not needed

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_description.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+## Hello World
```
Comment (Itay4) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_description.md:1]: update this

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml (hunk 1 above)
Comment (orlichter1) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:23]: going to implement this with @yaakovi tomorrow morning.

**Comment by orlichter1** [comment]
@Arsenikr Could you have a look at [this](https://github.com/demisto/content/blob/bdd8d305ef4ada4da381b593f2a4a579c53ea3cc/Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions_description.md) file and see if there are any more steps in order to enable this integration?

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    for i in parameters.split(','):
+        temp = i.strip()
+        if len(temp.split(':')) != 2:
+            raise ValueError(f"Key:Value pair {i} is not in the correct format.")
+        key, value = temp.split(':')
+        final_s += key.strip() + ':' + value.strip() + ','
+    final_s = final_s[:-1]
+    final_s = final_s.replace(',', '","')
+    final_s = '{"' + final_s.replace(':', '":"') + '"}'
+    return final_s
```
Comment (yaakovi) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:142]: what if value includes `"`
could be simplified:
image

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    region = args.get('region')
+    project_id = args.get('project_id')
+    res = client.functions_list(region, project_id)
+    functions = res.get('functions', [])
+    keys = list(functions[0].keys())
+    keys.remove('name')
+    disp_region = 'All' if client.region == '-' else client.region
+    hr = tableToMarkdown(f'Functions in project "{client.project}" and region "{disp_region}"',
+                         functions, ['name'] + keys)
+    ec = {'GoogleCloudFunctions.Function(val.name && val.name == obj.name)': functions}
```
Comment (yaakovi) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:92]: is a function name unique across all regions/projects? if not, add another key to the DT

Comment (orlichter1) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:92]: yes. The function's name is its full hierarchy: projects/gcp-integrations/locations/us-central1/functions/arseny-func

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/Pipfile [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+[dev-packages]
+pylint = "*"
+pytest = "==5.0.1"
+pytest-mock = "*"
+requests-mock = "*"
+pytest-asyncio = "*"
+
+[packages]
+pytest = "*"
+requests = "*"
```
[the remark this answers is not in the document]
Comment (orlichter1) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/Pipfile:15]: @yaakovi  I should have here all imports I'm doing? even though I have a docker image that supports it?

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml [hunk 10]
```diff
@@ -0,0 +1,166 @@
+category: Utilities
+commonfields:
+  id: Google Cloud Functions
+  version: -1
+configuration:
+- display: Service Account Private Key file contents (JSON)
+  name: credentials_json
+  required: true
+  type: 0
```
[the remark this answers is not in the document]
Comment (orlichter1) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:9]: @yaakovi What does the encrypted type do?

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    credentials_json = json.loads(demisto.params().get('credentials_json', {}))
+    project = demisto.params().get('project_id', '')
+    proxy = demisto.params().get('proxy', False)
+    region = demisto.params().get('region')
+    client = GoogleClient('cloudfunctions', 'v1', credentials_json, proxy, project=project, region=region)
+    commands = {
+        'google-cloud-functions-list': functions_list_command,
+        'google-cloud-function-regions-list': region_list_command,
+        'google-cloud-function-get-by-name': get_function_by_name_command,
+        'google-cloud-function-execute': execute_function_command
```
[the remark this answers is not in the document]
Comment (orlichter1) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:155]: @yaakovi This is the last element in the dictionary. Shouldn't it be without a comma at the end?

[the remark this answers is not in the document]
Comment (yaakovi) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:161]: when adding another element later, will prevent changing that line.
see pep8: https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.yml:9]: visible string vs dots-string

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/Pipfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/Pipfile:15]: list of all packages as in the dockerfiles repo (you can take the file from there)

**Comment by orlichter1** [comment]
@kirbles19 Can you please have a look at the mentioned docs? thanks

File: Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py [hunk 12]
```diff
@@ -151,14 +172,18 @@ def format_parameters(parameters: str) -> str:
 def main():
     credentials_json = json.loads(demisto.params().get('credentials_json', {}))
     project = demisto.params().get('project_id', '')
-    proxy = demisto.params().get('proxy', False)
     region = demisto.params().get('region')
-    client = GoogleClient('cloudfunctions', 'v1', credentials_json, proxy, project=project, region=region)
+    # proxy and insecure not yet implemented
```
Comment (yaakovi) [inline Packs/GoogleCloudFunctions/Integrations/GoogleCloudFunctions/GoogleCloudFunctions.py:176]: ?

**Comment by kirbles19** [comment]
@orlichter1 Done

**Pull request merged**